File: internal/agentrun/activity.go
package agentrun

// ActivityState is the small, content-free state vocabulary used by unified
// agent activity views. Unknown is retained explicitly so a successful
// process exit cannot be mistaken for objective completion.
type ActivityState string

const (
	ActivityRunning   ActivityState = "running"
	ActivityWaiting   ActivityState = "waiting"
	ActivityBlocked   ActivityState = "blocked"
	ActivityCompleted ActivityState = "completed"
	ActivityFailed    ActivityState = "failed"
	ActivityStale     ActivityState = "stale"
	ActivityUnknown   ActivityState = "unknown"
)

// DeriveActivityState combines lifecycle and structured intelligence without
// introducing a second inference model. Lifecycle staleness is authoritative;
// waiting and blocked require intelligence; active runs otherwise stay
// running. Completion and failure require the explicit evidence or process
// failure semantics already enforced by DeriveRunIntelligence.
func DeriveActivityState(run Run, intelligence RunIntelligence) ActivityState {
	if run.State == StateStale {
		return ActivityStale
	}
	active := runIsActive(run)
	switch intelligence.State {
	case IntelligenceFailed:
		return ActivityFailed
	case IntelligenceCompleted:
		return ActivityCompleted
	case IntelligenceBlocked:
		if active {
			return ActivityBlocked
		}
	case IntelligenceWaiting:
		if active {
			return ActivityWaiting
		}
	}
	if active {
		return ActivityRunning
	}
	return ActivityUnknown
}

File: crates/ptrack-agent/src/activity.rs
use serde::{Deserialize, Serialize};

use crate::run::run_is_active;
use crate::{IntelligenceState, Run, RunIntelligence, RunState};

#[derive(Clone, Copy, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub enum ActivityState {
    #[serde(rename = "running")]
    Running,
    #[serde(rename = "waiting")]
    Waiting,
    #[serde(rename = "blocked")]
    Blocked,
    #[serde(rename = "completed")]
    Completed,
    #[serde(rename = "failed")]
    Failed,
    #[serde(rename = "stale")]
    Stale,
    #[default]
    #[serde(rename = "unknown")]
    Unknown,
}

#[must_use]
pub fn derive_activity_state(run: &Run, intelligence: &RunIntelligence) -> ActivityState {
    if run.state == RunState::Stale {
        return ActivityState::Stale;
    }
    let active = run_is_active(run);
    match intelligence.state {
        IntelligenceState::Failed => ActivityState::Failed,
        IntelligenceState::Completed => ActivityState::Completed,
        IntelligenceState::Blocked if active => ActivityState::Blocked,
        IntelligenceState::Waiting if active => ActivityState::Waiting,
        _ if active => ActivityState::Running,
        _ => ActivityState::Unknown,
    }
}
